charmhelpers/fetch/ubuntu.py

```diff
@@ -189,17 +190,185 @@
 CMD_RETRY_COUNT = 3  # Retry a failing fatal command X times.
 
 
+def dpkg_list(packages):
+    """Get data from system dpkg database for package.
+
+    Note that one of the main purposes for this function is to avoid dependency
+    on the ``python-apt`` package.
+
+    The indicated package is a wrapper around the ``apt`` C++ library which is
+    tightly connected to the version of the distribution it is shipped on.
+    This in turn makes it incredibly hard to distribute as a wheel for a charm
+    that supports a large span of distro versions.
+
+    :param packages: Packages to get data from
+    :type packages: list[str]
+    :returns: Structured data about installed packages, keys like
+              ``dpkg-query --list``
+    :rtype: dict
+    :raises: subprocess.CalledProcessError
+    """
+    pkgs = {}
+    cmd = ['dpkg-query', '--list']
+    cmd.extend(packages)
+    if locale.getlocale() == (None, None):
+        # subprocess calls out to locale.getpreferredencoding(False) to
+        # determine encoding.  Workaround for Trusty where the
+        # environment appears to not be set up correctly.
+        locale.setlocale(locale.LC_ALL, 'en_US.UTF-8')
+    try:
+        output = subprocess.check_output(cmd,
+                                         stderr=subprocess.STDOUT,
+                                         universal_newlines=True)
+        headings = []
+        for line in output.splitlines():
+            if line.startswith('||/'):
+                headings = line.split()
+                headings.pop(0)
+                continue
+            elif (line.startswith('|') or line.startswith('+') or
+                  line.startswith('dpkg-query:')):
+                continue
+            data = line.split(None, 4)
+            status = data.pop(0)
+            if status != 'ii':
+                continue
+            pkg = {}
+            for kv in zip(headings, data):
+                pkg.update({kv[0].lower(): kv[1]})
+            if 'name' in pkg:
+                pkgs.update({pkg['name']: pkg})
+    except subprocess.CalledProcessError as cp:
+        # ``dpkg-query`` may return error and at the same time have produced
```

Comment:
Does this mean we *still* want to use the packages?  In which case, move the except up to just after the checkcall() and allow it to continue if it's the case;  here `pkgs` will be `{}`

Reply:
Good catch